server/src/main/java/org/elasticsearch/common/xcontent/XContentHelper.java

```diff
@@ -380,8 +380,13 @@ public static BytesReference toXContent(ToXContent toXContent, XContentType xCon
      */
     @Deprecated
     public static XContentType xContentType(BytesReference bytes) {
-        BytesRef br = bytes.toBytesRef();
-        return XContentFactory.xContentType(br.bytes, br.offset, br.length);
+        try {
+            final InputStream inputStream = bytes.streamInput();
+            return XContentFactory.xContentType(inputStream);
```

Comment:
We should probably catch and handle `IllegalArgumentException` in case the stream input doesn't support marking

Reply:
Right that's quite subtle, how about this: added an assertion for this explicitly since we rely on the fact that the `BytesReference` returns a stream that supports marking elsewhere as well and on second look also added the `BytesArray` fast-path to speed things up.
wdyt?

Reply:
Works for me. I do wonder if we should consider a separate change to BytesReference and add a `hasArray` method so we can stop using `instanceof BytesArray`